infra/src/main/kotlin/fi/fta/geoviite/infra/switchLibrary/SwitchStructureDao.kt

```diff
@@ -170,12 +183,42 @@ class SwitchStructureDao(jdbcTemplateParam: NamedParameterJdbcTemplate?) : DaoBa
             jdbcTemplate.queryForObject(sql, params) { rs, _ -> rs.getRowVersion("id", "version") }
                 ?: throw IllegalStateException("Failed to generate ID for new Location Track")
         insertJoints(version, switchStructure.joints)
-        insertAlignment(version, switchStructure.alignments)
+        insertAlignments(version, switchStructure.alignments)
 
         logger.daoAccess(AccessType.INSERT, SwitchStructure::class, version)
         return version
     }
 
+    @Transactional
+    fun updateSwitchStructure(switchStructure: SwitchStructure): RowVersion<SwitchStructure> {
+        val sql = """
+            update common.switch_structure
+            set
+                type = :type,
+                presentation_joint_number = :presentation_joint_number
+            where
+                id = :id
+            returning id, version
+        """.trimIndent()
+
+        val params = mapOf(
+            "id" to (switchStructure.id as IntId).intValue,
+            "type" to switchStructure.type.typeName,
+            "presentation_joint_number" to switchStructure.presentationJointNumber.intValue
+        )
+
+        jdbcTemplate.setUser()
+        val version: RowVersion<SwitchStructure> =
+            jdbcTemplate.queryForObject(sql, params) { rs, _ -> rs.getRowVersion("id", "version") }
+                ?: throw IllegalStateException("Failed to generate ID for new Location Track")
+        deleteJoints(version.id)
+        deleteAlignments(version.id)
+        insertJoints(version, switchStructure.joints)
```

Comment:
Näin tehtynä tuo ei säilytä alignmenttien/jointtien historiaa. Tai toki versiot kaikki säilyy, mutta päivittyvän jointin id vaihtuu joten versioiden välillä ei ole mitään yhteyttä. Tuo ei ehkä ole tässä tosin kovin tärkeää ja tämä pysyy mukavan yksinkertaisena näin. Oikeastaan nuo tarvitsisi jonkun luonnollisen avaimen jotta nuo palatkin voisi upsertata (ja poistotkin pitäisi silloin osata hoitaa). Eli OK näin, mutta jos tulee mieleen helppo tapa tunnistaa edellinen versio (esim jointilla jointnumber toimisi) niin tuo sisältökin voisi periaatteessa olla upsert. 

Käytännössä siis voi laittaa jonkun luonnollisen avaimen uniikiksi, jolloin insertistä voisi tehdä upsertin lisäämällä vain `on conflict (luonnollinen_avain) do update ...`. Toinen kaava jota käytin alignmentin segmentteihin layout-puolella oli että käsittelee niitä listana ja yhdisti segmentin avaimeksi alignment_id + segment_index, jolloin voi vain upsertata tuolla avaimella ja lopuksi sanoa isoimmalla indeksillä `delete ... where alignment_id = :a_id and segment_index > :max_index`. Tuossa toki oletetaan että elementtien järjestys listalla on fiksattu niin että siitä syntyy järkevä versiohistoria (mikä ei suoranaisesti pidä paikkaansa segmenteilläkään).